### PR TITLE
Added ability to upload code scanning to GitHub

### PR DESCRIPTION
**Pull request by omerzi**
- [x] All [tests](https://github.com/jfrog/froggit-go/actions/workflows/test.yml) passed. If this feature is not already covered by the tests, I added new tests.
- [x] I used `go fmt ./...` for formatting the code before submitting the pull request.
- [x] This feature is included on all supported VCS providers - GitHub, Bitbucket cloud, Bitbucket server, and GitLab.

---

### CODE REVIEW

File: vcsclient/github.go [hunk 1]
```diff
@@ -524,3 +568,32 @@ func mapGitHubPullRequestToPullRequestInfoList(pullRequestList []*github.PullReq
 	}
 	return
 }
+
+func convertToBase64(data []byte) string {
+	sEnc := b64.StdEncoding.EncodeToString(data)
```
Comment (talarian1) [inline vcsclient/github.go:573]:
```suggestion
	return b64.StdEncoding.EncodeToString(data)
```

File: vcsclient/github.go [hunk 2]
```diff
@@ -524,3 +568,32 @@ func mapGitHubPullRequestToPullRequestInfoList(pullRequestList []*github.PullReq
...
+	return sEnc
```
Comment (talarian1) [inline vcsclient/github.go:574]:
```suggestion
```

File: vcsclient/github.go [hunk 3]
```diff
@@ -524,3 +568,32 @@ func mapGitHubPullRequestToPullRequestInfoList(pullRequestList []*github.PullReq
...
+}
+
+func packScanningResult(data []byte) (string, error) {
+	compressedScan, err := compressPackage(data)
+	if err != nil {
+		return "", err
+	}
+	packScan := convertToBase64(compressedScan)
+
```
Comment (talarian1) [inline vcsclient/github.go:583]:
```suggestion
```

File: vcsclient/github.go [hunk 4]
```diff
@@ -524,3 +568,32 @@ func mapGitHubPullRequestToPullRequestInfoList(pullRequestList []*github.PullReq
...
+	sEnc := b64.StdEncoding.EncodeToString(data)
+	return sEnc
+}
+
+func packScanningResult(data []byte) (string, error) {
+	compressedScan, err := compressPackage(data)
+	if err != nil {
+		return "", err
+	}
+	packScan := convertToBase64(compressedScan)
```
Comment (talarian1) [inline vcsclient/github.go:582]:
```suggestion
```

File: vcsclient/github.go [hunk 5]
```diff
@@ -524,3 +568,32 @@ func mapGitHubPullRequestToPullRequestInfoList(pullRequestList []*github.PullReq
...
+
+	return packScan, nil
```
Comment (talarian1) [inline vcsclient/github.go:584]:
```suggestion
	return convertToBase64(compressedScan), nil
```

File: vcsclient/github.go [hunk 6]
```diff
@@ -524,3 +568,32 @@ func mapGitHubPullRequestToPullRequestInfoList(pullRequestList []*github.PullReq
...
+}
+
+func compressPackage(stream []byte) ([]byte, error) {
+	var b bytes.Buffer
+	w := gzip.NewWriter(&b)
+	_, err := w.Write(stream)
+	if err == nil {
```
Comment (talarian1) [inline vcsclient/github.go:591]:
```suggestion
	if err != nil {
```

File: vcsclient/github.go [hunk 7]
```diff
@@ -524,3 +568,32 @@ func mapGitHubPullRequestToPullRequestInfoList(pullRequestList []*github.PullReq
...
+		return "", err
+	}
+	packScan := convertToBase64(compressedScan)
+
+	return packScan, nil
+}
+
+func compressPackage(stream []byte) ([]byte, error) {
+	var b bytes.Buffer
+	w := gzip.NewWriter(&b)
```
Comment (talarian1) [inline vcsclient/github.go:589]: let's move the close statement to here (inside a differ call)

**Review by talarian1 — CHANGES_REQUESTED** [review]

File: vcsclient/github.go [hunk 8]
```diff
@@ -441,6 +444,47 @@ func (client *GitHubClient) UnlabelPullRequest(ctx context.Context, owner, repos
...
+	if err != nil {
+		return "", err
+	}
+	sarifID, resp, err := ghClient.CodeScanning.UploadSarif(ctx, owner, repository, &github.SarifAnalysis{
+		CommitSHA:   &commitSHA,
+		Ref:         &ref,
+		Sarif:       &packagedScan,
+		CheckoutURI: nil,
+	})
+	if err != nil && resp.Response.StatusCode != 202 {
```
Comment (talarian1) [inline vcsclient/github.go:469]: Let's just add an explanation here, regarding the 202 and the error mechanism.

File: vcsclient/bitbucketcloud.go [hunk 9]
```diff
@@ -440,6 +440,10 @@ func (client *BitbucketCloudClient) UnlabelPullRequest(ctx context.Context, owne
 	return errLabelsNotSupported
 }
 
+func (client *BitbucketCloudClient) UploadCodeScanning(ctx context.Context, owner string, repository string, branch string, scan string) (string, error) {
+	return "", errLabelsNotSupported
```
[the remark this answers is not in the document]
Comment (omerzi) [reply vcsclient/bitbucketcloud.go]: done

File: vcsclient/github.go (hunk 1 above)
Comment (omerzi) [reply vcsclient/github.go]: done

File: vcsclient/github.go (hunk 3 above)
Comment (omerzi) [reply vcsclient/github.go:584]: done

File: vcsclient/github.go (hunk 4 above)
Comment (omerzi) [reply vcsclient/github.go]: done

File: vcsclient/github.go (hunk 5 above)
Comment (omerzi) [reply vcsclient/github.go]: done

File: vcsclient/github.go (hunk 6 above)
Comment (omerzi) [reply vcsclient/github.go]: done

File: vcsclient/github.go (hunk 7 above)
Comment (omerzi) [reply vcsclient/github.go:590]: done

File: vcsclient/github.go (hunk 8 above)
Comment (omerzi) [reply vcsclient/github.go:471]: done

File: go.mod [hunk 10]
```diff
@@ -5,6 +5,7 @@ go 1.17
 require (
 	github.com/gfleury/go-bitbucket-v1 v0.0.0-20220418082332-711d7d5e805f
 	github.com/google/go-github/v41 v41.0.0
```
Comment (yahavi) [inline go.mod:7]: Let's use one of 41 or 45

File: vcsclient/gitlab.go [hunk 11]
```diff
@@ -374,6 +374,10 @@ func (client *GitLabClient) UnlabelPullRequest(ctx context.Context, owner, repos
 	return err
 }
 
+func (client *GitLabClient) UploadCodeScanning(ctx context.Context, owner string, repository string, branch string, scan string) (string, error) {
+	return "", errCodeScanningNotSupported
```
Comment (yahavi) [inline vcsclient/gitlab.go:378]: It looks like the error returned here is: "code scanning is not supported on Bitbucket"

File: vcsclient/vcsclient.go [hunk 12]
```diff
@@ -171,6 +171,13 @@ type VcsClient interface {
 	// name          - Label name
 	// pullRequestID - Pull request ID
 	UnlabelPullRequest(ctx context.Context, owner, repository, name string, pullRequestID int) error
+
+	// Upload Scanning Analysis uploads a scanning analysis file to the relevant git provider
+	// owner         - User or organization
+	// repository    - VCS repository name
+	// branch        - The name of the branch
+	// scan  		 - Code scanning analysis
+	UploadCodeScanning(ctx context.Context, owner, repository, branch, scan string) (string, error)
```
Comment (yahavi) [inline vcsclient/vcsclient.go:180]: Let's add this API to the README.md

**Review by yahavi — CHANGES_REQUESTED** [review]

File: go.mod (hunk 10 above)
Comment (omerzi) [reply go.mod]: done

File: vcsclient/gitlab.go (hunk 11 above)
Comment (omerzi) [reply vcsclient/gitlab.go]: done

File: vcsclient/vcsclient.go (hunk 12 above)
Comment (omerzi) [reply vcsclient/vcsclient.go:180]: done

File: README.md [hunk 13]
```diff
@@ -469,3 +469,23 @@ provider := vcsutils.GitHub
 
 webhookInfo, err := webhookparser.ParseIncomingWebhook(provider, token, request)
 ```
+
+### Upload Code Scanning
```
Comment (yahavi) [inline README.md:473]:
```suggestion
#### Upload Code Scanning
```

Additionally, please move this API above the Webhook parser

Comment (omerzi) [reply README.md]: Added

File: vcsclient/bitbucketserver.go [hunk 14]
```diff
@@ -598,3 +603,7 @@ func (client *BitbucketServerClient) mapBitbucketServerCommitToCommitInfo(commit
 		ParentHashes:  parents,
 	}
 }
+
+func (client *BitbucketServerClient) UploadCodeScanning(ctx context.Context, owner string, repository string, branch string, scan string) (string, error) {
```
Comment (yahavi) [inline vcsclient/bitbucketserver.go:607]: Please make sure to do that in all other implementations:
```suggestion
func (client *BitbucketServerClient) UploadCodeScanning(ctx context.Context, owner string, repository string, branch string, scanResults string) (string, error) {
```

**Review by yahavi — CHANGES_REQUESTED** [review]

File: vcsclient/bitbucketserver.go (hunk 14 above)
Comment (omerzi) [reply vcsclient/bitbucketserver.go]: Done

**Review by yahavi — APPROVED** [review]

**Pull request merged**